rfd/0141-tsh-bench-db.md

```diff
@@ -0,0 +1,279 @@
+databases (without Teleport proxying them). The team can use this to compare
+Teleport performance with direct database access.
+
+## Prior art
+
+### `tsh bench` subcommands
+The `tsh` already has commands for performing benchmarks. Those rely on the
+structure defined at `lib/benchmark/benchmark.go`. It defines a small framework
+to define benchmarks. By following it, we can rely on the tools already used on
+the SSH and Kubernetes benchmarks, such as reports.
+
+To help readers better understand how the database benchmark will be organized,
+here is a summary of it:
+* Benchmark suites: Implements a function that generates workload functions.
+  The suite is called before the benchmark starts and returns a workload function.
+  It has access to the Teleport Client.
+* Workload function: Functions that will be repeatedly executed to grab data.
+  Since the suite generates it, it can use information from the suite function.
+  For example, on the SSH benchmark, the servers list used to select the server
+  where the command will be executed is fetched on the suite function, and the
+  workload function only connects to it.
+
+#### Common options
+The benchmark structure already supports common flags to parametrize the test
+execution:
+- `--rate`: Control how many workload functions are executed per second. For
+  example, it can control how many connections will be sent to the target
+  database per second.
+- `--duration`: Defines for how long the test will be executed.
+
+#### Results
+The benchmark commands follow a standardized result output. There is also an
+option to export the test results to a more detailed version using the `--export`
+flag.
+
+```code
+$ tsh bench ...
+* Requests originated: 9
+* Requests failed: 0
+
+Histogram
+
+Percentile Response Duration
+---------- -----------------
+25         123 ms
+50         125 ms
+75         129 ms
+90         130 ms
+95         130 ms
+99         130 ms
+100        130 ms
+```
+
+### [`mysqlslap`](https://dev.mysql.com/doc/refman/5.7/en/mysqlslap.html)
+
+> `mysqlslap` is a diagnostic program designed to emulate client load for a
+MySQL server and to report the timing of each stage. It works as if multiple
+clients are accessing the server.
+
+It has to working formats:
+* Auto-generates database tables and queries. When used like this, it has flags
+  to customize table columns (set how many columns there will be) and choose the
+  type of queries it will execute during the benchmark (for example, `read` will
+  only execute SELECT queries).
+* Custom queries. It has a parameter to setup a table and insert necessary data.
+  And executes the custom query provided by the user.
+
+## Details
+The database subcommands will be divided per protocol. This gives more
+flexibility for each protocol to require their arguments and flags.
+
+The "root" command for each protocol will execute the following:
+* Starts a connection with the database (using local proxy);
```

Comment:
Are we going to start a single local proxy or multiple? I wonder if we should have a mode that can simulate either a single client connecting through a proxy and doing many requests, or multiple clients (each of which uses a separate proxy).

Reply:
The initial idea was to start a single proxy to avoid consuming too many resources from the client (like open ports). But we can keep this a default behavior and add a new flag to start a single proxy per connection. Should we do something similar to certificate generation?